File: app/src/main/java/fr/corenting/edcompanion/network/NewsNetwork.java
package fr.corenting.edcompanion.network;

import android.content.Context;

import org.greenrobot.eventbus.EventBus;

import java.util.ArrayList;
import java.util.List;

import fr.corenting.edcompanion.models.NewsArticle;
import fr.corenting.edcompanion.models.apis.EDAPIV4.NewsArticleResponse;
import fr.corenting.edcompanion.models.events.GalnetNews;
import fr.corenting.edcompanion.models.events.News;
import fr.corenting.edcompanion.network.retrofit.EDApiV4Retrofit;
import fr.corenting.edcompanion.singletons.RetrofitSingleton;
import retrofit2.Call;
import retrofit2.internal.EverythingIsNonNull;

public class NewsNetwork {
    public static void getGalnetNews(Context ctx, String language) {

        // Set default language if none provided
        if (language == null || language.equals("")) {
            language = "en";
        }

        EDApiV4Retrofit retrofit = RetrofitSingleton.getInstance()
                .getEdApiV4Retrofit(ctx.getApplicationContext());

        retrofit2.Callback<List<NewsArticleResponse>> callback = new retrofit2.Callback<List<NewsArticleResponse>>() {
            @Override
            @EverythingIsNonNull
            public void onResponse(Call<List<NewsArticleResponse>> call,
                                   retrofit2.Response<List<NewsArticleResponse>> response) {
                List<NewsArticleResponse> body = response.body();
                if (!response.isSuccessful() || body == null) {
                    onFailure(call, new Exception("Invalid response"));
                } else {
                    GalnetNews news;
                    try {
                        List<NewsArticle> articles = new ArrayList<>();
                        for (NewsArticleResponse item : body) {
                            articles.add(NewsArticle.Companion.fromNewsArticleResponse(item));
                        }
                        news = new GalnetNews(true, articles);
                    } catch (Exception e) {
                        news = new GalnetNews(false, new ArrayList<>());
                    }
                    EventBus.getDefault().post(news);
                }
            }

            @Override
            @EverythingIsNonNull
            public void onFailure(Call<List<NewsArticleResponse>> call,
                                  Throwable t) {
                GalnetNews news = new GalnetNews(false, new ArrayList<>());
                EventBus.getDefault().post(news);
            }
        };
        retrofit.getGalnetNews(language).enqueue(callback);
    }

    public static void getNews(Context ctx, String language) {

        // Set default language if none provided
        if (language == null || language.equals("")) {
            language = "en";
        }
        EDApiV4Retrofit retrofit = RetrofitSingleton.getInstance()
                .getEdApiV4Retrofit(ctx.getApplicationContext());

        retrofit2.Callback<List<NewsArticleResponse>> callback = new retrofit2.Callback<List<NewsArticleResponse>>() {
            @Override
            @EverythingIsNonNull
            public void onResponse(Call<List<NewsArticleResponse>> call,
                                   retrofit2.Response<List<NewsArticleResponse>> response) {
                List<NewsArticleResponse> body = response.body();
                if (!response.isSuccessful() || body == null) {
                    onFailure(call, new Exception("Invalid response"));
                } else {
                    News news;
                    try {
                        List<NewsArticle> articles = new ArrayList<>();
                        for (NewsArticleResponse item : body) {
                            articles.add(NewsArticle.Companion.fromNewsArticleResponse(item));
                        }
                        news = new News(true, articles);
                    } catch (Exception e) {
                        news = new News(false, new ArrayList<>());
                    }
                    EventBus.getDefault().post(news);
                }
            }

            @Override
            @EverythingIsNonNull
            public void onFailure(Call<List<NewsArticleResponse>> call,
                                  Throwable t) {
                News news = new News(false, new ArrayList<>());
                EventBus.getDefault().post(news);
            }
        };
        retrofit.getNews(language).enqueue(callback);
    }
}

File: app/src/main/java/fr/corenting/edcompanion/network/NewsNetwork.kt
package fr.corenting.edcompanion.network

import android.content.Context
import fr.corenting.edcompanion.models.NewsArticle
import fr.corenting.edcompanion.models.NewsArticle.Companion.fromNewsArticleResponse
import fr.corenting.edcompanion.models.ProxyResult
import fr.corenting.edcompanion.models.events.News
import fr.corenting.edcompanion.singletons.RetrofitSingleton

object NewsNetwork {

    suspend fun getNews(ctx: Context, initialLanguage: String): ProxyResult<News> {
        // Set default language if none provided
        var language = initialLanguage
        if (language == "") {
            language = "en"
        }
        val retrofit = RetrofitSingleton.getInstance()
            .getEdApiV4Retrofit(ctx.applicationContext)

        try {
            val newsRet = retrofit.getNews(language)

            val news = try {
                val articles: MutableList<NewsArticle> = ArrayList()
                for (item in newsRet) {
                    articles.add(fromNewsArticleResponse(item))
                }
                News(articles)
            } catch (e: Exception) {
                News(ArrayList())
            }
            return ProxyResult(news, null)
        } catch (t: Throwable) {
            return ProxyResult(null, t)
        }
    }

    suspend fun getGalnetNews(ctx: Context, initialLanguage: String): ProxyResult<News> {
        // Set default language if none provided
        var language = initialLanguage
        if (language == "") {
            language = "en"
        }
        val retrofit = RetrofitSingleton.getInstance()
            .getEdApiV4Retrofit(ctx.applicationContext)

        try {
            val newsRet = retrofit.getGalnetNews(language)

            val news = try {
                val articles: MutableList<NewsArticle> = ArrayList()
                for (item in newsRet) {
                    articles.add(fromNewsArticleResponse(item))
                }
                News(articles)
            } catch (e: Exception) {
                News(ArrayList())
            }
            return ProxyResult(news, null)
        } catch (t: Throwable) {
            return ProxyResult(null, t)
        }
    }
}